libraries/AC_PrecLand/AC_PrecLand_StateMachine.cpp

```diff
@@ -0,0 +1,254 @@
+#include "AC_PrecLand_StateMachine.h"
+
+static const float MAX_POS_ERROR_CM = 75.0f;  // Maximum possition error for retry locations
+static const uint32_t FAILSAFE_INIT_TIMEOUT_MS = 7000;   // Timeout in ms before failsafe measures are started. During this period vehicle is completely stopped to give user the time to take over
+static const float RETRY_OFFSET_ALT_CM = 150.0f;  // This gets added to the altitude of the retry location
+
+// Constructor
+AC_PrecLand_StateMachine::AC_PrecLand_StateMachine(const AP_InertialNav& inav) :
+    _inertial_nav(inav)
+{
+}
+
+// Initialize the state machine. This is called everytime vehicle switches mode
+void AC_PrecLand_StateMachine::init()
+{
+    // init is only called ONCE per mode change. So in a particuar mode we can retry only a finite times.
+    // The counter will be reset if the statemachine is called from a different mode
+    _retry_count = 0;
+    // reset every other statemachine
+    reset_failed_landing_statemachines();
+}
+
+// Reset the landing statemachines. This needs to be called everytime the landing target is back in sight.
+// So that if the landing target goes out of sight again, we can start the failed landing procedure back from the beginning stage
+void AC_PrecLand_StateMachine::reset_failed_landing_statemachines()
```

Comment:
maybe remove the "s" at the end?  I think there's is only one state machine.. there are multiple bits of state but there is only one state machine..